custom_components/adaptive_lighting/switch.py

```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
     return service_datas
 
 
-async def handle_apply(switch: AdaptiveSwitch, service_call: ServiceCall):
-    """Handle the entity service apply."""
-    hass = switch.hass
-    data = service_call.data
-    all_lights = data[CONF_LIGHTS]
-    if not all_lights:
-        all_lights = switch._lights
-    all_lights = _expand_light_groups(hass, all_lights)
-    switch.turn_on_off_listener.lights.update(all_lights)
+def _parse_service_args(hass: HomeAssistant, service_call: ServiceCall):
     _LOGGER.debug(
-        "Called 'adaptive_lighting.apply' service with '%s'",
-        data,
+        "Called '_parse_service_args' with service data:\n'%s'",
+        service_call.data,
     )
-    for light in all_lights:
-        if data[CONF_TURN_ON_LIGHTS] or is_on(hass, light):
-            await switch._adapt_light(  # pylint: disable=protected-access
-                light,
-                data[CONF_TRANSITION],
-                data[ATTR_ADAPT_BRIGHTNESS],
-                data[ATTR_ADAPT_COLOR],
-                data[CONF_PREFER_RGB_COLOR],
-                force=True,
-                context=switch.create_context("service", parent=service_call.context),
+    data = service_call.data
+    lights = data[CONF_LIGHTS]
+    switch_unique_id = data.get("entity_id")
+    switch = None
+    if not lights and not switch_unique_id:
+        # alternatively just loop and call the service for every switch.
+        _LOGGER.debug(
+            "adaptive-lighting: If you were trying to adapt every single light on every single"
+            " switch, please inform the developers at"
+            " https://github.com/basnijholt/adaptive-lighting of your use case."
+            " Currently you must pass either the adapt main switch, or the lights to every"
+            " service call. adaptive-lighting will now raise a soft error for you in the meantime."
+        )
+        _LOGGER.error(
+            "bad service data to adaptive-lighting service call - "
+            "you must pass either an entity_id or a light. Service data:\n%s",
+            service_call.data,
+        )
+        raise ValueError(
+            "adaptive-lighting: No switch or light was passed to service call."
+        )
+
+    if switch_unique_id is not None:
+        if len(switch_unique_id) > 1 and lights:
+            _LOGGER.error(
+                "bad service data: cannot pass multiple switch entities while also passing"
+                "any lights. Got %s",
+                service_call.data,
+            )
+            raise ValueError(
+                "adaptive-lighting: Multiple switches passed with lights argument"
             )
+        switch = []
+        for unique_id in switch_unique_id:
+            ent_reg = entity_registry.async_get(hass)
+            ent_entry = ent_reg.async_get(unique_id)
+            # `AttributeError: 'NoneType' object has no attribute 'config_entry_id'` fix later maybe
```

Comment:
When does this exception occur?

Reply:
It's resolved, that's an old reminder to myself during tests when the sequence existed as:

```
        ent_reg = entity_registry.async_get(hass)
        ent_entry = ent_reg.async_get(str(data['entity_id']))
        config_id = ent_entry.config_entry_id
        switch = hass.data[DOMAIN][config_id]["instance"]
```

Reply:
see https://github.com/basnijholt/adaptive-lighting/pull/459/commits/aeed1da0b561e3c2b0aead1d02aab90f930afb7f